### PR TITLE
Implement spilling for PartialSortExec

### LINKED ISSUE #9153: Expand PartialSort usage to bounded input cases to reduce memory usage
### Is your feature request related to a problem or challenge?

PartialSort proposed in https://github.com/apache/arrow-datafusion/pull/9125 can also be used on bounded input as well which will help reducing memory usage for the sorts when the input is already partially sorted (input plan ordering and plan required ordering has a common prefix)

### Describe the solution you'd like

We might need to consider incorporating `ExternalSorter` in `PartialSortExec` plan as it is used in `SortExec`. Then it should be completely safe to replace `SortExec` with `PartialSortExec` in `EnforceSorting` without checking for unboundedness of input.

### Describe alternatives you've considered

Implementing partial sort logic inside `SortExec` and making `SortExec` choose which sorting logic to run might be an option.

It can also be argued that it is easier to understand the intention when `PartialSortExec` and `SortExec` are kept as separate plans. 

### Additional context

https://github.com/apache/arrow-datafusion/issues/7456

### PR DESCRIPTION
**Pull request by yyy1000**
## Which issue does this PR close?

Hopefully it will
Closes #9153 #9170.

## Rationale for this change

See the issues

## What changes are included in this PR?

This refactor wants to use `ExternalSorter` to implement `PartialSort`.
`ExternalSorter` has implement `spilling`

## Are these changes tested?

Yes, a new test case for spilling

## Are there any user-facing changes?

No

### CODE REVIEW

File: datafusion/physical-plan/src/sorts/partial_sort.rs [hunk 1]
```diff
@@ -510,6 +398,62 @@ mod tests {
         Ok(())
     }
 
+    #[tokio::test]
+    async fn test_partial_sort_spill() -> Result<()> {
```
Comment (yyy1000) [inline datafusion/physical-plan/src/sorts/partial_sort.rs:402]: This is the new test case for spill, I mainly copy from sort.rs

File: datafusion/physical-plan/src/sorts/partial_sort.rs [hunk 2]
```diff
@@ -563,7 +507,7 @@ mod tests {
                 "| 1 | 3 | 0 |",
                 "+---+---+---+",
             ];
-            assert_eq!(2, result.len());
+            assert_eq!(1, result.len());
```
Comment (yyy1000) [inline datafusion/physical-plan/src/sorts/partial_sort.rs:510]: I think the sort in ExternalSort will merge the two `RecordBatch` to one if no spilling?

File: datafusion/physical-plan/src/sorts/partial_sort.rs [hunk 3]
```diff
@@ -647,157 +591,6 @@ mod tests {
         Ok(())
     }
 
-    fn prepare_partitioned_input() -> Arc<dyn ExecutionPlan> {
```
Comment (yyy1000) [inline datafusion/physical-plan/src/sorts/partial_sort.rs:650]: I deleted these test case for the same reason, the sort in ExternalSort will merge the RecordBatchs so I think these test cases are not necessary?

File: datafusion/physical-plan/src/sorts/sort.rs [hunk 4]
```diff
@@ -309,6 +310,66 @@ impl ExternalSorter {
...
+        if let Some(slice_point) = self.get_slice_point(prefix, &input)? {
+            self.in_mem_batches.push(input.slice(0, slice_point));
+            self.in_mem_batches
+                .push(input.slice(slice_point, input.num_rows() - slice_point));
+        } else {
+            self.in_mem_batches.push(input);
+        }
+        self.in_mem_batches_sorted = false;
+        Ok(())
+    }
```
Comment (yyy1000) [inline datafusion/physical-plan/src/sorts/sort.rs:372]: Here it also first `get_slice_point` and then push them in `in_mem_batches`.
A question I have is, in the past, the 2nd RecordBatch will concat to the remaining of the 1st RecordBatch, but here I didn't implement that. 🤔 Don't know whether it's necessary.

**Review by yyy1000 — COMMENTED** [review]
Hi @alamb 
I tried to help the PartialSortExec for spilling and maybe general usage case.
I have some questions below and hope you can review when you are available.

**Comment by yyy1000** [comment]
The CI fails because PartialSort will replace Sort in some test cases like below, I can change that if others in this PR looks good.
```
expected:

[
    "SortPreservingMergeExec: [a@0 ASC,b@1 ASC]",
    "  SortExec: expr=[a@0 ASC,b@1 ASC,c@2 ASC]",
    "    RepartitionExec: partitioning=RoundRobinBatch(10), input_partitions=1",
    "      CsvExec: file_groups={1 group: [[x]]}, projection=[a, b, c, d, e], output_ordering=[a@0 ASC, b@1 ASC], has_header=false",
]
actual:

[
    "SortPreservingMergeExec: [a@0 ASC,b@1 ASC]",
    "  PartialSortExec: expr=[a@0 ASC,b@1 ASC,c@2 ASC], common_prefix_length=[2]",
    "    RepartitionExec: partitioning=RoundRobinBatch(10), input_partitions=1",
    "      CsvExec: file_groups={1 group: [[x]]}, projection=[a, b, c, d, e], output_ordering=[a@0 ASC, b@1 ASC], has_header=false",
]
```

File: datafusion/physical-plan/src/sorts/partial_sort.rs [hunk 5]
```diff
@@ -282,16 +276,31 @@ impl ExecutionPlan for PartialSortExec {
...
+        );
+        let prefix = self.common_prefix_length;
+        Ok(Box::pin(RecordBatchStreamAdapter::new(
+            self.schema(),
+            futures::stream::once(async move {
+                while let Some(batch) = input.next().await {
+                    let batch = batch?;
+                    sorter.insert_batch_with_prefix(batch, prefix).await?;
+                }
+                sorter.sort()
```
Comment (mustafasrepo) [inline datafusion/physical-plan/src/sorts/partial_sort.rs:300]: Our original aim in the `PartialSortExec` was to produce streaming results when existing ordering and required ordering has a common prefix ordering. If that is the case, `PartialSortExec` can produce result chunk by chunk (e.g not as a single bulk). In this code snippet data is fed to chunk by chunk. However, `.sort` is called once, at the end after `while` loop ends. For this reason, data is produced still produced as a single bulk.

I think this is the reason why existing test fails in this implementation

Comment (yyy1000) [reply datafusion/physical-plan/src/sorts/partial_sort.rs:300]: Got it. Thanks!
Do you have some suggestions? I think maybe calling `sort_in_mem_batches` in the original implementation after `insert_batch_with_prefix` each time can solve this, and add the spilling function to `sort_in_mem_batches`. 🤔

Comment (yyy1000) [reply datafusion/physical-plan/src/sorts/partial_sort.rs:300]: Also I have a question is `sort` returns a `SendableRecordBatchStream` and produce bulk, and `PartialSort`  returns chuck by chunk. What're the pros of `chuck by chunk`? I think they do the same for memory usage. 🤔

Comment (mustafasrepo) [reply datafusion/physical-plan/src/sorts/partial_sort.rs:300]: > Got it. Thanks! Do you have some suggestions? I think maybe calling `sort_in_mem_batches` in the original implementation after `insert_batch_with_prefix` each time can solve this, and add the spilling function to `sort_in_mem_batches`. 🤔

Not really sure. How to implement it. Unfortunately, I am not super familiar with `ExternalSort` code.

Comment (yyy1000) [reply datafusion/physical-plan/src/sorts/partial_sort.rs:300]: Yeah, thanks for your info.
I think in this PR, PartialSort can also reduce memory usage. It also emits data chunk by chunk but return a merge stream at last. Let's see what others view this.

**Comment by alamb** [comment]
Does anyone have any actual queries running against real inputs that use PartialSort but also run out of memory?

I agree it can happen in theory, but I was wondering if anyone has seen it happening in the real world

**Comment by yyy1000** [comment]
> Does anyone have any actual queries running against real inputs that use PartialSort but also run out of memory?

No, I also don't have one. I just test the spilling function using the test from Sort. 👀

**Comment by mustafasrepo** [comment]
> Does anyone have any actual queries running against real inputs that use PartialSort but also run out of memory?
> 
> I agree it can happen in theory, but I was wondering if anyone has seen it happening in the real world

I also think, it is quite unlikely.

**Pull request closed without merging**